libs/core/langchain_core/prompts/chat.py

```diff
@@ -921,6 +921,86 @@ class ChatPromptTemplate(BaseChatPromptTemplate):
+                  A message can be represented using the following formats:
+                  (1) BaseMessagePromptTemplate, (2) BaseMessage, (3) 2-tuple of
+                  (message type, template); e.g., ("human", "{user_input}"),
+                  (4) 2-tuple of (message class, template), (4) a string which is
+                  shorthand for ("human", template); e.g., "{user_input}".
+            template_format: format of the template. Defaults to "f-string".
+            input_variables: A list of the names of the variables whose values are
+                required as inputs to the prompt.
+            optional_variables: A list of the names of the variables that are optional
+                in the prompt.
+            partial_variables: A dictionary of the partial variables the prompt
+                template carries. Partial variables populate the template so that you
+                don't need to pass them in every time you call the prompt.
+            validate_template: Whether to validate the template.
+            input_types: A dictionary of the types of the variables the prompt template
+                expects. If not provided, all variables are assumed to be strings.
+
+        Returns:
+            A chat prompt template.
+
+        Examples:
+
+            Instantiation from a list of message templates:
+
+            .. code-block:: python
+
+                template = ChatPromptTemplate([
+                    ("human", "Hello, how are you?"),
+                    ("ai", "I'm doing well, thanks!"),
+                    ("human", "That's good to hear."),
+                ])
+
+            Instantiation from mixed message formats:
+
+            .. code-block:: python
+
+                template = ChatPromptTemplate([
+                    SystemMessage(content="hello"),
+                    ("human", "Hello, how are you?"),
+                ])
+
+        """
+        _messages = [
+            _convert_to_message(message, template_format) for message in messages
+        ]
+
+        # Automatically infer input variables from messages
+        input_vars: Set[str] = set()
+        optional_variables: Set[str] = set()
+        partial_vars: Dict[str, Any] = {}
+        for _message in _messages:
+            if isinstance(_message, MessagesPlaceholder) and _message.optional:
+                partial_vars[_message.variable_name] = []
+                optional_variables.add(_message.variable_name)
+            elif isinstance(
+                _message, (BaseChatPromptTemplate, BaseMessagePromptTemplate)
+            ):
+                input_vars.update(_message.input_variables)
+
+        kwargs = {
+            **dict(
+                input_variables=sorted(input_vars),
+                optional_variables=sorted(optional_variables),
+                partial_variables=partial_vars,
+            ),
+            **kwargs,
+        }
+        cast(Type[ChatPromptTemplate], super()).__init__(messages=_messages, **kwargs)
```

Comment:
Why is cast needed

Reply:
mypy thinks its ``super().__init__()`` output BaseChatPromptTemplate